### ISSUE TITLE
What is the correct way to deal with shared modules?

### ISSUE THREAD

**Issue by aalapargekartibco** [TIBCOSoftware/bw6-plugin-maven#543]
I am able to find information about this in bits and pieces but nothing which explains the complete process from design time to run time.

We are using a separate github repository per BW application and a separate one for shared modules. So if I have 10 BW applications then I have 11 github repositories, 1 per app and 1 for all the shared modules.

In design time, I imported one application and the required shared modules into my workspace and generated the pom. When I did that for the next application, the path of the parent pom in the shared module got updated.  This is problem #1.

In Jenkins, I have a one job per application and one for the shared modules, so 11 in total. So each application and the shared modules get cloned into the respective workspaces. So the relative path of the shard modules with respect to the parent pom on jenkins server can be different from on my local system. This is problem #2.

The bottom line question is - What is the step by step process from generating pom to creating jenkins build jobs for applications and shared modules that reside on separate github repositories ?

**Comment by vpawar-ai** [comment]
Hi @aalapargekartibco The error - `[ERROR] Caused by: java.lang.IllegalArgumentException: Can not set org.eclipse.aether.spi.log.Logger field org.apache.maven.repository.internal.DefaultVersionResolver.logger to org.eclipse.aether.internal.impl.slf4j.Slf4jLoggerFactory
` should get resolved once you add maven 3.6.x in maven installations via eclipse preferences. Please ensure that you select the added maven installation as default. The steps mentioned looks fine.

**Comment by aalapargekartibco** [comment]
Hi @vpawar-ai 

![image](https://user-images.githubusercontent.com/20302957/102914027-ad900700-444d-11eb-93be-3dd78bc26f74.png)
As can be seen, I have selected the maven installation.  I am still getting the error. At the bottom I see this message "Note: Embedded runtime is always used for dependency resolution". Is that kinda overriding the maven selection ?

You said "Please ensure that you select the added maven installation as default". How to do that ? I don't see any option that says "set as default" or something like that.

I tried restarting studio, made sure maven is selcted and tried, no luck.

**Comment by aalapargekartibco** [comment]
@vpawar-ai  I did this:
- Imported shared module into workspace , generated pom and installed. I see the shared module jar in my local .m2.
- Deleted everything from workspace and imported application.
- right click application module to add dependency, don't see that option. For the "add dependency" option to come, there should be pom.xml under application module. So the "generate pom for application" by right clicking application has to be done first. To do that either the shared module or its binary should be imported into workspace. So I imported the binary and generated pom for application.
- Now I am able to add dependency in the application module.
- the pom.xml under module now shows the shared module dependency.
- but the pom.xml under parent points to the actual shared module using relative path under the "modules" tab because of the "generate pom for application" step.

- On Jenkins, the build failed because the parent pom.xml is looking for the actual shared module using relative path.

Can you please tell me what am I missing?

**Comment by aalapargekartibco** [comment]
@vpawar-ai  I removed the shared module reference manually from the parent pom. And it worked.
Is there a way to prevent the shared module reference from appearing in the parent pom in the first place?

**Comment by vpawar-ai** [comment]
@aalapargekartibco There are 2 ways to use shared modules -
1. Shared Module with Maven (Use as child module in the parent project) - The Shared module project should be present in the workspace and the POM generation wizard will pick it, mavenize the project and generates pom.xml. The shared module is included as child module in parent pom.xml.

2. External Shared Modules with maven - The external shared module is added as pom dependency in application module. The steps you should follow -
-  Create Shared Module Project. Use "Generate POM for Shared Module" option.
- Execute `mvn clean install` to install the shared module jar into local m2 repository.
- Create new eclipse workspace. Import your application project. (Do not import shared module project)
- Launch "Generate POM for Application" wizard and mavenize the application project.
- Open application module's "pom.xml" and add pom dependency(as explained at - https://github.com/TIBCOSoftware/bw6-plugin-maven/wiki/SharedModule#step-5). Save the changes. This should import external shared module project in "read only mode" in your workspace.
- The External shared module will not be added as module in parent's pom.xml. It will be used as pom dependency in application module.

**Comment by aalapargekartibco** [comment]
@vpawar-ai  These steps that you mentioned are not working for me:
_Create new eclipse workspace. Import your application project. (Do not import shared module project)
Launch "Generate POM for Application" wizard and mavenize the application project._

When I try to generate pom I get this error 
![error](https://user-images.githubusercontent.com/20302957/103102808-ce8e5e80-45eb-11eb-9623-d45f2e531649.jpg)

It is not letting me generate the pom without importing the shared module

**Comment by vpawar-ai** [comment]
@aalapargekartibco You are seeing this error because the shared module project is added in the `<Your-application-project>->Package Unit->Includes`. So the POM generation wizard expects it to be present in the workspace. Please remove it from the includes list and add it as POM dependency as explained in the steps.
Please note the difference between using Shared module as maven child Module vs adding External Shared Module as POM dependency.

**Comment by aalapargekartibco** [comment]
@vpawar-ai  I removed the shared module from the "includes" section of the application and generated the pom. I am now getting this error in the project. 

Description	Resource	Path	Location	Type
Project build error: Non-readable POM C:\git\Tibco-SalesForce\SalesForce\SalesForce.application.parent\pom.xml: input contained no data	pom.xml	/SalesForce	line 1	Maven pom Loading Problem

Description	Resource	Path	Location	Type
Project build error: Non-readable POM C:\git\Tibco-SalesForce\SalesForce\SalesForce.application.parent\pom.xml: input contained no data	pom.xml	/SalesForce.application	line 1	Maven pom Loading Problem

These are the values that I provided while generating pom. 

![generate pom](https://user-images.githubusercontent.com/20302957/103138679-f4c30580-46a2-11eb-8b08-204d3bc25538.jpg)

Is there anything in here that needs to be mentioned and I did not mention?
